source/d3d8types.cpp

```diff
@@ -98,40 +94,65 @@ void convert_surface_desc(D3DSURFACE_DESC &input, D3DSURFACE_DESC8 &output)
 	output.Usage = input.Usage;
 	output.Pool = input.Pool;
 	output.Size = calc_texture_size(input.Width, input.Height, 1, input.Format);
-	output.MultiSampleType = input.MultiSampleType;
 	output.Width = input.Width;
 	output.Height = input.Height;
 
-	if (output.MultiSampleType == D3DMULTISAMPLE_NONMASKABLE)
+	// Check for D3DMULTISAMPLE_NONMASKABLE and change it to D3DMULTISAMPLE_NONE for best D3D8 compatibility.
+	if (input.MultiSampleType == D3DMULTISAMPLE_NONMASKABLE)
 	{
 		output.MultiSampleType = D3DMULTISAMPLE_NONE;
 	}
+	else
+	{
+		output.MultiSampleType = input.MultiSampleType;
+	}
 }
 void convert_present_parameters(D3DPRESENT_PARAMETERS8 &input, D3DPRESENT_PARAMETERS &output)
 {
 	output.BackBufferWidth = input.BackBufferWidth;
 	output.BackBufferHeight = input.BackBufferHeight;
 	output.BackBufferFormat = input.BackBufferFormat;
 	output.BackBufferCount = input.BackBufferCount;
-	output.MultiSampleType = input.MultiSampleType;
 	output.MultiSampleQuality = 0;
-	output.SwapEffect = input.SwapEffect;
 	output.hDeviceWindow = input.hDeviceWindow;
 	output.Windowed = input.Windowed;
 	output.EnableAutoDepthStencil = input.EnableAutoDepthStencil;
 	output.AutoDepthStencilFormat = input.AutoDepthStencilFormat;
 	output.Flags = input.Flags;
 	output.FullScreen_RefreshRateInHz = input.FullScreen_RefreshRateInHz;
-	output.PresentationInterval = input.FullScreen_PresentationInterval;
 
-	if (output.SwapEffect == D3DSWAPEFFECT_COPY_VSYNC)
+	// MultiSampleType must be D3DMULTISAMPLE_NONE unless SwapEffect has been set to D3DSWAPEFFECT_DISCARD.
+	// https://msdn.microsoft.com/en-us/library/windows/desktop/bb172588(v=vs.85).aspx
+	// Check for D3DMULTISAMPLE_NONMASKABLE and change it to D3DMULTISAMPLE_NONE for best D3D8 compatibility.
+	if (input.SwapEffect != D3DSWAPEFFECT_DISCARD || input.MultiSampleType == D3DMULTISAMPLE_NONMASKABLE)
 	{
-		output.SwapEffect = D3DSWAPEFFECT_COPY;
+		output.MultiSampleType = D3DMULTISAMPLE_NONE;
+	}
+	else
```

Comment:
I rather prefer avoiding the else statements here by moving that code in front of the if as it was before. Other than that it looks great.